### PR TITLE
[AGENT-5408] Integration with moderations library

### PR DESCRIPTION
**Pull request by amarmudrankit**
This is the first step to integrate moderations with DRUM

# This repository is public. Do not put here any private DataRobot or customer's data: code, datasets, model artifacts, .etc.

### CODE REVIEW

**Comment by datarobot-devinfragit3** [comment]
pipelineController framework is deprecated, please migrate CICD for this repository to Harness

**Comment by datarobot-devinfragit3** [comment]
pipelineController framework is deprecated, please migrate CICD for this repository to Harness

**Comment by datarobot-devinfragit3** [comment]
pipelineController framework is deprecated, please migrate CICD for this repository to Harness

**Comment by amarmudrankit** [comment]
The wrapper code would look like follows:
```
# DRUM_SCORE_WRAPPER=".moderate/score_wrapper.py"
# def score_wrapper(score_func, mode, df):
#     # Read config
#     # call prompt moderation
#
#     # call customer score
#
#     # call response moderations
#
#     # return result
#     pass
#
# score_wrapper(customer_score, model, df)
```

File: custom_model_runner/datarobot_drum/drum/adapters/model_adapters/python_model_adapter.py [hunk 1]
```diff
@@ -560,13 +603,33 @@ def _predict_legacy_drum(self, data, model, **kwargs) -> RawPredictResponse:
...
         extra_model_output = None
         if self._custom_hooks.get(CustomHooks.SCORE):
-            try:
-                # noinspection PyCallingNonCallable
-                predictions_df = self._custom_hooks.get(CustomHooks.SCORE)(data, model, **kwargs)
-            except Exception as exc:
-                self._log_and_raise_final_error(
-                    exc, "Model 'score' hook failed to make predictions."
-                )
+            if self._target_type == TargetType.TEXT_GENERATION and self._guard_pipeline:
```
Comment (yakov-g) [inline custom_model_runner/datarobot_drum/drum/adapters/model_adapters/python_model_adapter.py:606]: I think you cam move if clause inside try/except and have only one _log_and_raise_final_error

File: custom_model_runner/datarobot_drum/drum/adapters/model_adapters/python_model_adapter.py [hunk 2]
```diff
@@ -560,13 +603,33 @@ def _predict_legacy_drum(self, data, model, **kwargs) -> RawPredictResponse:
...
+                try:
+                    predictions_df = self._guard_moderation_hooks[GUARD_SCORE_WRAPPER_NAME](
+                        data,
+                        model,
+                        self._guard_pipeline,
+                        self._custom_hooks.get(CustomHooks.SCORE),
```
Comment (yakov-g) [inline custom_model_runner/datarobot_drum/drum/adapters/model_adapters/python_model_adapter.py:612]: Does every text generation model has `score` hook?

File: custom_model_runner/datarobot_drum/drum/adapters/model_adapters/python_model_adapter.py [hunk 3]
```diff
@@ -560,13 +603,33 @@ def _predict_legacy_drum(self, data, model, **kwargs) -> RawPredictResponse:
...
-            except Exception as exc:
-                self._log_and_raise_final_error(
-                    exc, "Model 'score' hook failed to make predictions."
-                )
+            if self._target_type == TargetType.TEXT_GENERATION and self._guard_pipeline:
+                try:
+                    predictions_df = self._guard_moderation_hooks[GUARD_SCORE_WRAPPER_NAME](
+                        data,
+                        model,
+                        self._guard_pipeline,
```
Comment (yakov-g) [inline custom_model_runner/datarobot_drum/drum/adapters/model_adapters/python_model_adapter.py:611]: Do you pass init hook back to wrapper hook?

File: custom_model_runner/datarobot_drum/drum/adapters/model_adapters/python_model_adapter.py [hunk 4]
```diff
@@ -103,14 +105,55 @@ def __init__(self, model_dir, target_type=None):
...
+            guard_module = __import__(guard_file_prefix)
+            self._guard_moderation_hooks = {
+                GUARD_INIT_HOOK_NAME: getattr(guard_module, GUARD_INIT_HOOK_NAME, None),
+                GUARD_SCORE_WRAPPER_NAME: getattr(guard_module, GUARD_SCORE_WRAPPER_NAME, None),
+            }
+            if (
+                self._guard_moderation_hooks[GUARD_INIT_HOOK_NAME]
+                and self._guard_moderation_hooks[GUARD_SCORE_WRAPPER_NAME]
+            ):
+                self._guard_pipeline = self._guard_moderation_hooks[GUARD_INIT_HOOK_NAME]()
```
Comment (yakov-g) [inline custom_model_runner/datarobot_drum/drum/adapters/model_adapters/python_model_adapter.py:149]: Why var is called `_guard_pipeline` if you assign an `init` hook?

File: custom_model_runner/datarobot_drum/drum/adapters/model_adapters/python_model_adapter.py [hunk 5]
```diff
@@ -103,14 +105,55 @@ def __init__(self, model_dir, target_type=None):
...
                 raise ValueError("Unexpected empty target name for text generation!")
+            self._load_guard_hooks_for_drum()
         else:
             self._target_name = None
 
+    def _load_guard_hooks_for_drum(self):
+        if os.environ.get("MLOPS_GUARD_HOOK_FILE", None) is None:
+            return
+
+        guard_file_prefix = os.environ.get("MLOPS_GUARD_HOOK_FILE")
```
Comment (yakov-g) [inline custom_model_runner/datarobot_drum/drum/adapters/model_adapters/python_model_adapter.py:123]: Could you please do some refactoring and unify common code for the regular hooks loading and guards hook loading?

**Review by yakov-g — CHANGES_REQUESTED** [review]
1. Please add tests
2. I added pylint check to the testing pipelines, but made it to never fail - there are so much errors. I would like to ask everyone to help with cleaning the code. If you have some time, please run pylint locally on the files you change and fix some of the issues.

File: custom_model_runner/datarobot_drum/drum/adapters/model_adapters/python_model_adapter.py (hunk 4 above)
Comment (amarmudrankit) [reply custom_model_runner/datarobot_drum/drum/adapters/model_adapters/python_model_adapter.py:149]: If you notice carefully, it is not `init` hook being assigned, but `init` hook being "called" "()" at the end.  The init hook will return the the `pipeline` object.

File: custom_model_runner/datarobot_drum/drum/adapters/model_adapters/python_model_adapter.py (hunk 3 above)
Comment (amarmudrankit) [reply custom_model_runner/datarobot_drum/drum/adapters/model_adapters/python_model_adapter.py:611]: As mentioned below, init hook will return the pipeline object, which is passed to the score wrapper.

File: custom_model_runner/datarobot_drum/drum/adapters/model_adapters/python_model_adapter.py (hunk 5 above)
Comment (amarmudrankit) [reply custom_model_runner/datarobot_drum/drum/adapters/model_adapters/python_model_adapter.py:123]: I will try to see if I can, didn't want to break the stuff which is already running.

**Comment by amarmudrankit** [comment]
I will definitely add the tests.
Regarding black changes, I can help you with the changes in this file, but in a separate PR, otherwise it will be harder to review

**Review by yakov-g — APPROVED** [review]

**Comment by amarmudrankit** [comment]
jenkins run test_functional_r_environment

**Review by rvorobii — APPROVED** [review]

**Comment by amarmudrankit** [comment]
jenkins run test_functional_r_environment

**Comment by amarmudrankit** [comment]
jenkins run test_functional_general

**Comment by amarmudrankit** [comment]
@datarobot/core-modeling can you take a look at the PR

**Review by akshoop — APPROVED** [review]
LGTM from Core Modeling, CC @eric-s-s @scottp-ml for awareness

**Pull request merged**